{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import modules and relevant files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "importing modules  \n"
     ]
    }
   ],
   "source": [
    "print 'importing modules  '\n",
    "import os\n",
    "import subprocess\n",
    "from datetime import datetime as time\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import Counter\n",
    "from Bio.Seq import Seq\n",
    "%matplotlib inline\n",
    "\n",
    "### UPDATE\n",
    "sequence_file = \"./.fastq.gz\"\n",
    "\n",
    "sequence_file_name = \"./\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extracting primer sequences and barcodes from the reference files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "116\n",
      "AGGTCTCTATGAACTTGCGCTCAATTCCCGT\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "#CCATCTCATCCCTGCGTGTCTCCGACTCAG ATGACTCAGT GATGTCCACGAGGTCTCT\n",
    "#                    TCCGACTCAG ATGACTCAGT GATGTCCACGAGGTCTCT   [20:]\n",
    "#                    TCCGACTCAG ATGACTCAGT GATGTCCACG           [20:50]\n",
    "#CCATCTCATCCCTGCGTGTCTCCGACTCAG ATAGCAGTGC GATGTCCACGAGGTCTCT\n",
    "#CCATCTCATCCCTGCGTGTCTCCGACTCAG ACGATCTATC GATGTCCACGAGGTCTCT\n",
    "#CCATCTCATCCCTGCGTGTCTCCGACTCAG AGCATGTCAT GATGTCCACGAGGTCTCT\n",
    "#CCATCTCATCCCTGCGTGTCTCCGACTCAG CGAGTATGAT GATGTCCACGAGGTCTCT\n",
    "\n",
    "#                               ACGATCTATC GATGTCCACGAGGTCTCTATGAACTTGCGCTCAATTCCCGTATGCTGCAGGTCGACGGATCCCCGGGTTAATTAAGGCGCGCCAGATCTGTTTAGCTATAGTGTGC\n",
    "#                                                                                   GCGCCAGATCTGTTTAGCACTGAGTCATATCACCGACTGCCCATAGAGAGGAAAGCGG\n",
    "\n",
    "#ATGAACTTGCGCTCAATTCCCGTATGCTGCAGGTCGACGGATCCCCGGGTTAATTAAGGCGCGCCAGATCTGTTTAGC TATAGTGTGC\n",
    "#                                                            GCGCCAGATCTGTTTAGC ACTGAGTCAT ATCACCGACTGCCCATAGAGAGGAAAGCGG\n",
    "#                                                            GCGCCAGATCTGTTTAGC GAGCTATATC ATCACCGACTGCCCATAGAGAGGAAAGCGG\n",
    "#                                                                                 GCTATATC ATCACCGACTGCCCATAGAGAG  [20:50]\n",
    "#                                                                      TGTTTAGC GAGCTATATC ATCACCGACTGC            [10:40]\n",
    "#                                                                    TCTGTTTAGC GAGCTATATC ATCACCGACT              [8:38]\n",
    "# ACGATCTATC                                                                                                  TATAGTGTGC\n",
    "# ACGATCTATC GATGTCCACGAGGTCTCT ATGAACTTGCGCTCAATTCC CGTATGCTGCAGGTCGACGGATCCCCGGGTTAATTAAGGCGCGCCAGATCTGTTTAGC TATAGTGTGC\n",
    "#                               ATGAACTTGCGCTCAATTCC\n",
    "# \n",
    "\n",
    "#print Seq(index_seq_dict['7|21|1|D1|A9'][1]).reverse_complement()[8:38]\n",
    "\n",
    "test_seq = 'ACGATCTATCGATGTCCACGAGGTCTCTATGAACTTGCGCTCAATTCCCGTATGCTGCAGGTCGACGGATCCCCGGGTTAATTAAGGCGCGCCAGATCTGTTTAGCTATAGTGTGC'\n",
    "\n",
    "strain_barcode_up_flank = 'CGAGGTCTCT'\n",
    "\n",
    "strain_barcode_down_flank = 'CGTATGCTGC'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "###########################  CGAGGTCTCTATGAACTTGCGCTCAATTCCCGTATGCTGC\n",
    "print len('ACGATCTATCGATGTCCACGAGGTCTCTATGAACTTGCGCTCAATTCCCGTATGCTGCAGGTCGACGGATCCCCGGGTTAATTAAGGCGCGCCAGATCTGTTTAGCTATAGTGTGC')\n",
    "# min read len will be 116\n",
    "print 'ACGATCTATCGATGTCCACGAGGTCTCTATGAACTTGCGCTCAATTCCCGTATGCTGCAGGTCGACGGATCCCCGGGTTAATTAAGGCGCGCCAGATCTGTTTAGCTATAGTGTGC'[20:51]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(('C8', 'A7'), '2|16|1'), (('C9', 'A4'), '3|5|2'), (('D4', 'A8'), '10|16|2'), (('D2', 'A10'), '8|21|2'), (('C10', 'A6'), '4|10|2'), (('C11', 'A4'), '5|5|2'), (('D5', 'A5'), '11|10|1'), (('D4', 'A5'), '10|10|1'), (('C10', 'A9'), '4|21|1'), (('D5', 'A8'), '11|16|2'), (('C12', 'A8'), '6|16|2'), (('C7', 'A4'), '1|5|2'), (('C8', 'A4'), '2|5|2'), (('C9', 'A7'), '3|16|1'), (('C10', 'A3'), '4|5|1'), (('C11', 'A7'), '5|16|1'), (('C11', 'A10'), '5|21|2'), (('C12', 'A5'), '6|10|1'), (('D7', 'A4'), '12|5|2'), (('C7', 'A10'), '1|21|2'), (('D7', 'A7'), '12|16|1'), (('D3', 'A5'), '9|10|1'), (('C8', 'A10'), '2|21|2'), (('D8', 'A12'), '2|0|1'), (('D1', 'A8'), '7|16|2'), (('C12', 'A6'), '6|10|2'), (('D7', 'A10'), '12|21|2'), (('D2', 'A7'), '8|16|1'), (('D3', 'A6'), '9|10|2'), (('D1', 'A6'), '7|10|2'), (('C12', 'A3'), '6|5|1'), (('C8', 'A8'), '2|16|2'), (('C9', 'A3'), '3|5|1'), (('C7', 'A9'), '1|21|1'), (('C10', 'A7'), '4|16|1'), (('C11', 'A3'), '5|5|1'), (('D5', 'A6'), '11|10|2'), (('D4', 'A4'), '10|5|2'), (('D5', 'A9'), '11|21|1'), (('C12', 'A9'), '6|21|1'), (('C7', 'A7'), '1|16|1'), (('D7', 'A8'), '12|16|2'), (('C8', 'A5'), '2|10|1'), (('D2', 'A9'), '8|21|1'), (('C9', 'A6'), '3|10|2'), (('C11', 'A9'), '5|21|1'), (('C10', 'A4'), '4|5|2'), (('C11', 'A6'), '5|10|2'), (('D5', 'A3'), '11|5|1'), (('D4', 'A7'), '10|16|1'), (('D4', 'A10'), '10|21|2'), (('D7', 'A5'), '12|10|1'), (('D2', 'A3'), '8|5|1'), (('C9', 'A9'), '3|21|1'), (('D8', 'A11'), '1|0|1'), (('C10', 'A10'), '4|21|2'), (('C12', 'A7'), '6|16|1'), (('D7', 'A6'), '12|10|2'), (('D2', 'A6'), '8|10|2'), (('D1', 'A10'), '7|21|2'), (('D9', 'A12'), '2|0|2'), (('D3', 'A7'), '9|16|1'), (('D1', 'A7'), '7|16|1'), (('D9', 'A11'), '1|0|2'), (('C7', 'A3'), '1|5|1'), (('D7', 'A3'), '12|5|1'), (('C8', 'A9'), '2|21|1'), (('D2', 'A5'), '8|10|1'), (('D1', 'A9'), '7|21|1'), (('D3', 'A8'), '9|16|2'), (('D1', 'A4'), '7|5|2'), (('D5', 'A7'), '11|16|1'), (('D4', 'A3'), '10|5|1'), (('C7', 'A6'), '1|10|2'), (('C8', 'A6'), '2|10|2'), (('D2', 'A8'), '8|16|2'), (('C9', 'A5'), '3|10|1'), (('C11', 'A8'), '5|16|2'), (('D4', 'A9'), '10|21|1'), (('C10', 'A5'), '4|10|1'), (('C11', 'A5'), '5|10|1'), (('D5', 'A4'), '11|5|2'), (('D4', 'A6'), '10|10|2'), (('C10', 'A8'), '4|16|2'), (('C7', 'A5'), '1|10|1'), (('C8', 'A3'), '2|5|1'), (('D3', 'A3'), '9|5|1'), (('D1', 'A3'), '7|5|1'), (('D5', 'A10'), '11|21|2'), (('C12', 'A4'), '6|5|2'), (('C7', 'A8'), '1|16|2'), (('D3', 'A4'), '9|5|2'), (('D7', 'A9'), '12|21|1'), (('C9', 'A8'), '3|16|2'), (('C12', 'A10'), '6|21|2'), (('C9', 'A10'), '3|21|2'), (('D2', 'A4'), '8|5|2'), (('D3', 'A9'), '9|21|1'), (('D1', 'A5'), '7|10|1'), (('D3', 'A10'), '9|21|2')]\n",
      "{'D5': 'TCCGACTCAGCGAGTATGATGATGTCCACG', 'D2': 'TCCGACTCAGATGACTCAGTGATGTCCACG', 'D9': 'TCCGACTCAGCTGTAGTCAGGATGTCCACG', 'D4': 'TCCGACTCAGCGAGATCTCTGATGTCCACG', 'C9': 'TCCGACTCAGAGCATGTCATGATGTCCACG', 'C8': 'TCCGACTCAGACGATCTATCGATGTCCACG', 'D8': 'TCCGACTCAGCTACTCACTAGATGTCCACG', 'D7': 'TCCGACTCAGCATAGCTATCGATGTCCACG', 'C12': 'TCCGACTCAGATACTGACGCGATGTCCACG', 'C11': 'TCCGACTCAGAGTCTGTACGGATGTCCACG', 'C10': 'TCCGACTCAGAGCTGACACAGATGTCCACG', 'C7': 'TCCGACTCAGAGATAGTGCTGATGTCCACG', 'D3': 'TCCGACTCAGCGTCATGATAGATGTCCACG', 'D1': 'TCCGACTCAGATAGCAGTGCGATGTCCACG'}\n",
      "{'A11': 'TCTGTTTAGCACGCAGCATCATCACCGACT', 'A10': 'TCTGTTTAGCATCAGAGATCATCACCGACT', 'A12': 'TCTGTTTAGCCTGTCACATCATCACCGACT', 'A3': 'TCTGTTTAGCCTGCATCTCTATCACCGACT', 'A5': 'TCTGTTTAGCACTGAGTCATATCACCGACT', 'A4': 'TCTGTTTAGCATCAGCACGTATCACCGACT', 'A7': 'TCTGTTTAGCTATCATGACGATCACCGACT', 'A6': 'TCTGTTTAGCACGATGTACGATCACCGACT', 'A9': 'TCTGTTTAGCGAGCTATATCATCACCGACT', 'A8': 'TCTGTTTAGCCGACGAGTAGATCACCGACT'}\n"
     ]
    }
   ],
   "source": [
    "indexes_for_libraries = './Barcodes/indexes_for_libraries.csv'\n",
    "\n",
    "\n",
    "index_lib_dict = {}\n",
    "\n",
    "lib_index_by_tuple = {}\n",
    "\n",
    "for_index_seq ={}\n",
    "rev_index_seq ={}\n",
    "\n",
    "\n",
    "with open(indexes_for_libraries, 'r') as source:\n",
    "    \n",
    "    next(source)\n",
    "    \n",
    "    for line in source:\n",
    "        \n",
    "        line = line.strip('\\n')\n",
    "        line = line.split(',')\n",
    "        \n",
    "        sample = '|'.join([line[0], line[1], line[2]])\n",
    "        indexes = [line[3], line[5]]\n",
    "        \n",
    "        index_lib_dict[sample] = indexes\n",
    "        \n",
    "        if line[3] not in for_index_seq.keys():\n",
    "            \n",
    "            for_index_seq[line[3]] = line[4][20:50]\n",
    "            \n",
    "        if line[5] not in rev_index_seq.keys():\n",
    "            \n",
    "            rev_index_seq[line[5]] = str(Seq(line[6]).reverse_complement())[8:38]\n",
    "            \n",
    "            \n",
    "        lib_index_by_tuple[tuple(indexes)] = sample\n",
    "        \n",
    "        \n",
    "       \n",
    "        \n",
    "print lib_index_by_tuple.items()\n",
    "#looks good\n",
    "\n",
    "print for_index_seq\n",
    "print rev_index_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## For and Rev Indexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for_indexes = './for_indexes.fa'\n",
    "rev_indexes = './rev_indexes.fa'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "with open(for_indexes, 'w') as output:\n",
    "    \n",
    "    for f_index in for_index_seq.keys():\n",
    "        \n",
    "        fasta_header = '>'+ f_index + '\\n'\n",
    "        fasta_seq =  for_index_seq[f_index] + '\\n'\n",
    "\n",
    "        output.write(fasta_header)\n",
    "        output.write(fasta_seq)\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "with open(rev_indexes, 'w') as output:\n",
    "    \n",
    "    for r_index in rev_index_seq.keys():\n",
    "        \n",
    "        fasta_header = '>'+ r_index + '\\n'\n",
    "        fasta_seq =  rev_index_seq[r_index] + '\\n'\n",
    "\n",
    "        output.write(fasta_header)\n",
    "        output.write(fasta_seq)\n",
    "        \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## strain barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "strain_fasta = './strain_barcodes.fa'\n",
    "\n",
    "strain_bc_key = './Barcodes/Strains_pool_light_noNfmt.csv'\n",
    "\n",
    "with open(strain_fasta, 'w') as output:\n",
    "    \n",
    "    with open(strain_bc_key, 'r') as source:\n",
    "\n",
    "                next(source)\n",
    "\n",
    "                for line in source:\n",
    "\n",
    "                    line_data = line.strip('\\n').split(';')\n",
    "\n",
    "                    guide_name = '|'.join([line_data[0], line_data[1], line_data[2]])\n",
    "                    guide_seq = line_data[3]\n",
    "\n",
    "                    fasta_header = '>'+ guide_name + '\\n'\n",
    "                    fasta_seq = strain_barcode_up_flank + guide_seq + strain_barcode_down_flank +'\\n'\n",
    "\n",
    "                    output.write(fasta_header)\n",
    "                    output.write(fasta_seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-35-f0b4f949ed2a>, line 4)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-35-f0b4f949ed2a>\"\u001b[1;36m, line \u001b[1;32m4\u001b[0m\n\u001b[1;33m    bowtie-build -f -r -o 4 for_indexes.fa bleuven_ION_nov2018_for_indexes\u001b[0m\n\u001b[1;37m                          ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "# command line calls\n",
    "\n",
    "# for indexes\n",
    "bowtie-build -f -r -o 4 for_indexes.fa bleuven_ION_nov2018_for_indexes\n",
    "\n",
    "# rev indexes\n",
    "bowtie-build -f -r -o 4 rev_indexes.fa bleuven_ION_nov2018_rev_indexes\n",
    "\n",
    "# guides\n",
    "bowtie-build -f -r -o 4 strain_barcodes.fa bleuven_ION_nov2018_strain_indexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fastqc A.Dube_Cabsoil_p01_c01.fastq.gz\n",
    "\n",
    "java -jar trimmomatic-0.36.jar SE -threa 6 -phred33 /home/phdes74/path_to_Dropbox/lab_manuscripts/Bleuven_ION/A.Dube_Cabsoil_p01_c01.fastq.gz /home/phdes74/path_to_Dropbox/lab_manuscripts/Bleuven_ION/A.Dube_Cabsoil_p01_c01_trim_110.fastq.gz MINLEN:115 CROP:115\n",
    "      \n",
    "fastqc A.Dube_Cabsoil_p01_c01_trim_110.fastq.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bowtie -t -v 0 -p 6 -k 1 --trim3 97 --norc -u 40000000 bleuven_ION_nov2018_for_indexes A.Dube_Cabsoil_p01_c01_trim_110.fastq index_for_1.txt\n",
      "bowtie -t -v 0 -p 6 -k 1 --trim5 100 --norc -u 40000000 bleuven_ION_nov2018_rev_indexes A.Dube_Cabsoil_p01_c01_trim_110.fastq index_rev_1.txt\n",
      "bowtie -t -v 0 -p 6 -k 1 --trim3 70 --trim5 19 --norc -m 1 -u 40000000 bleuven_ION_nov2018_strain_indexes A.Dube_Cabsoil_p01_c01_trim_110.fastq strain_bc_1.txt\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "index_for_align_param = 'bowtie -t -v 0 -p 6 -k 1 --trim3 97 --norc -u 40000000 bleuven_ION_nov2018_for_indexes '\n",
    "# parameters for forward index alignments. Based on optimization work done in BAR-SEQ notebook\n",
    "index_for_align_output = 'index_for_1.txt'\n",
    "# alignment output filepath as string\n",
    "index_for_align_call = index_for_align_param + 'A.Dube_Cabsoil_p01_c01_trim_110.fastq' + ' ' + index_for_align_output\n",
    "\n",
    "print index_for_align_call\n",
    "\n",
    "# alignement call as a string\n",
    "subprocess.check_output(index_for_align_call, shell = True)\n",
    "# performs the alignment\n",
    "\n",
    "\n",
    "\n",
    "index_rev_align_param = 'bowtie -t -v 0 -p 6 -k 1 --trim5 100 --norc -u 40000000 bleuven_ION_nov2018_rev_indexes '\n",
    "# parameters for reverse index alignments. Based on optimization work done in BAR-SEQ notebook\n",
    "index_rev_align_output = 'index_rev_1.txt'\n",
    "# alignment output filepath as string\n",
    "index_rev_align_call = index_rev_align_param + 'A.Dube_Cabsoil_p01_c01_trim_110.fastq' + ' ' + index_rev_align_output\n",
    "# alignement call as a string\n",
    "subprocess.check_output(index_rev_align_call, shell = True)\n",
    "# performs the alignment\n",
    "\n",
    "print index_rev_align_call\n",
    "\n",
    "\n",
    "\n",
    "guide_align_param = 'bowtie -t -v 0 -p 6 -k 1 --trim3 70 --trim5 19 --norc -m 1 -u 40000000 bleuven_ION_nov2018_strain_indexes '\n",
    "# parameters for perfect guide alignments. Based on optimization work done in BAR-SEQ notebook\n",
    "bc_align_output = 'strain_bc_1.txt'\n",
    "# alignment output filepath as string\n",
    "bc_align_call = guide_align_param + 'A.Dube_Cabsoil_p01_c01_trim_110.fastq' + ' ' + bc_align_output\n",
    "# alignement call as a string\n",
    "subprocess.check_output(bc_align_call, shell = True)\n",
    "# performs the alignment\n",
    "\n",
    "print bc_align_call"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bowtie -t -v 0 -p 6 -k 1 --trim3 70 --trim5 19 --norc -m 1 -u 40000000 bleuven_ION_nov2018_strain_indexes A.Dube_Cabsoil_p01_c01_trim_110.fastq strain_bc_1.txt\n"
     ]
    }
   ],
   "source": [
    "print bc_align_call"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "guide_counts_counter_dict = {}\n",
    "# a dict that will hold one instance of a Counter() object for each library in the experiment (20 total). Each of\n",
    "# those Counter objects will be used the count the number of reads for each guide in each library based the bowtie\n",
    "# alignment data\n",
    "\n",
    "\n",
    "for lib in index_lib_dict.keys():\n",
    "    # loops through the samples in the experiment\n",
    "    \n",
    "    guide_counts_counter_dict[lib] = Counter()\n",
    "    # creates an empty Counter() for guide count\n",
    "\n",
    "bc_align_output = 'strain_bc_1.txt'\n",
    "index_for_align_output = 'index_for_1.txt'\n",
    "index_rev_align_output = 'index_rev_1.txt'\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38344726\n",
      "35515990\n",
      "34085271\n",
      "35678281\n"
     ]
    }
   ],
   "source": [
    "index_found_for = {}\n",
    "index_found_rev = {}\n",
    "# empty containers that will hold    read_name: index    data pairs for the forward and reverse reads that\n",
    "# had valid alignments\n",
    "\n",
    "\n",
    "with open(index_for_align_output, 'r') as for_index_positives:\n",
    "    for line in for_index_positives:\n",
    "        # opens P1 index alignment pack file and loops through lines\n",
    "\n",
    "        line = line.split('\\t')\n",
    "        # split line to extract data\n",
    "        read_name = line[0]\n",
    "        # Only keep the cluster coordinates and run associated info, not the illumina annotations in the read \n",
    "        # name. See http://support.illumina.com/content/dam/illumina-support/help/BaseSpaceHelp_v2/Content/Vault/\n",
    "        #           Informatics/Sequencing_Analysis/BS/swSEQ_mBS_FASTQFiles.htm\n",
    "        # For info on illumina read name convention\n",
    "        index_found_for[read_name] = line[2]\n",
    "        # add read P1 index to dict\n",
    "\n",
    "print len(index_found_for.keys())\n",
    "# log entry: number of reads with valid P1 index alignments\n",
    "\n",
    "with open(index_rev_align_output, 'r') as rev_index_positives:\n",
    "    for line in rev_index_positives:\n",
    "        # opens P2 index alignment pack file and loops through lines\n",
    "\n",
    "        line = line.split('\\t')\n",
    "        # split line to extract data\n",
    "        read_name = line[0]\n",
    "        # Only keep the cluster coordinates and run associated info\n",
    "        index_found_rev[read_name] = line[2]\n",
    "        # add read P2 index to dict\n",
    "\n",
    "print len(index_found_rev.keys())\n",
    "# log entry: number of reads with valid P2 index alignments\n",
    "\n",
    "\n",
    "both_indexes = set(index_found_for.keys()) & set(index_found_rev.keys())\n",
    "print len(both_indexes)\n",
    "# Use the set data format to quickly id reads with valid alignments for both P1 and P2 indexes, and stores the\n",
    "# taht number for the read pack in the log\n",
    "\n",
    "del both_indexes\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "barcode_match = {}\n",
    "# empty containers that will hold    read_name: guide    data pairs for the forward reads that had alignments\n",
    "# with 0 mismatches at the right location in the reference sequence set\n",
    "\n",
    "with open(bc_align_output, 'r') as matches:\n",
    "    for line in matches:\n",
    "        # opens gRNA alignment pack file and loops through lines\n",
    "\n",
    "        line = line.split('\\t')\n",
    "        # split line to extract data\n",
    "        read_name = line[0]\n",
    "        # Only keep the cluster coordinates and run associated info\n",
    "        guide = line[2]\n",
    "        # the guide for which a match was found\n",
    "\n",
    "        if int(line[3]) in range(0, 5):\n",
    "            # the alignment is at the right location in the guide sequence\n",
    "\n",
    "            barcode_match[read_name] = guide\n",
    "            # add read:guide to dict\n",
    "\n",
    "print len(barcode_match.keys())\n",
    "# adds the number of perfect matches found in the pack of reads\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30812839\n"
     ]
    }
   ],
   "source": [
    "indexes_and_guides = set(index_found_for.keys()) & set(index_found_rev.keys()) & set(barcode_match.keys())\n",
    "print len(indexes_and_guides)\n",
    "# uses sets in python to find the set of reads for which the P1, P2 and gRNA alignments were good\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('C8', 'A7'), ('C9', 'A4'), ('D4', 'A8'), ('D2', 'A10'), ('C10', 'A6'), ('C11', 'A4'), ('D5', 'A5'), ('D4', 'A5'), ('C10', 'A9'), ('D5', 'A8'), ('C12', 'A8'), ('C7', 'A4'), ('C8', 'A4'), ('C9', 'A7'), ('C10', 'A3'), ('C11', 'A7'), ('C11', 'A10'), ('C12', 'A5'), ('D7', 'A4'), ('C7', 'A10'), ('D7', 'A7'), ('D3', 'A5'), ('C8', 'A10'), ('D8', 'A12'), ('D1', 'A8'), ('C12', 'A6'), ('D7', 'A10'), ('D2', 'A7'), ('D3', 'A6'), ('D1', 'A6'), ('C12', 'A3'), ('C8', 'A8'), ('C9', 'A3'), ('C7', 'A9'), ('C10', 'A7'), ('C11', 'A3'), ('D5', 'A6'), ('D4', 'A4'), ('D5', 'A9'), ('C12', 'A9'), ('C7', 'A7'), ('D7', 'A8'), ('C8', 'A5'), ('D2', 'A9'), ('C9', 'A6'), ('C11', 'A9'), ('C10', 'A4'), ('C11', 'A6'), ('D5', 'A3'), ('D4', 'A7'), ('D4', 'A10'), ('D7', 'A5'), ('D2', 'A3'), ('C9', 'A9'), ('D8', 'A11'), ('C10', 'A10'), ('C12', 'A7'), ('D7', 'A6'), ('D2', 'A6'), ('D1', 'A10'), ('D9', 'A12'), ('D3', 'A7'), ('D1', 'A7'), ('D9', 'A11'), ('C7', 'A3'), ('D7', 'A3'), ('C8', 'A9'), ('D2', 'A5'), ('D1', 'A9'), ('D3', 'A8'), ('D1', 'A4'), ('D5', 'A7'), ('D4', 'A3'), ('C7', 'A6'), ('C8', 'A6'), ('D2', 'A8'), ('C9', 'A5'), ('C11', 'A8'), ('D4', 'A9'), ('C10', 'A5'), ('C11', 'A5'), ('D5', 'A4'), ('D4', 'A6'), ('C10', 'A8'), ('C7', 'A5'), ('C8', 'A3'), ('D3', 'A3'), ('D1', 'A3'), ('D5', 'A10'), ('C12', 'A4'), ('C7', 'A8'), ('D3', 'A4'), ('D7', 'A9'), ('C9', 'A8'), ('C12', 'A10'), ('C9', 'A10'), ('D2', 'A4'), ('D3', 'A9'), ('D1', 'A5'), ('D3', 'A10')]\n"
     ]
    }
   ],
   "source": [
    "print lib_index_by_tuple.keys()\n",
    "\n",
    "correct_libs = set([x for x in lib_index_by_tuple.keys()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1737\n"
     ]
    }
   ],
   "source": [
    "chimera = 0\n",
    "\n",
    "for read in indexes_and_guides:\n",
    "    # loops through the reads tht passed the filter\n",
    "\n",
    "    for_i = index_found_for[read]\n",
    "    # gets the P1 index of the read\n",
    "\n",
    "    rev_i = index_found_rev[read]\n",
    "    # gets the P2 index of the read\n",
    "    \n",
    "    if tuple((for_i, rev_i)) in correct_libs:\n",
    "\n",
    "        lib = lib_index_by_tuple[tuple((for_i, rev_i))]\n",
    "        # gets the sample to which the read should be assigned based on P1, P2 combination\n",
    "\n",
    "        guide = barcode_match[read]\n",
    "        # gets the guide associated with the read\n",
    "\n",
    "        guide_counts_counter_dict[lib][guide] += 1\n",
    "        # a count for the guide detected in its specific sample\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        chimera += 1 \n",
    "\n",
    "\n",
    "\n",
    "del index_found_for\n",
    "del index_found_rev\n",
    "del barcode_match\n",
    "\n",
    "print chimera"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## for part 2 of reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bowtie -t -v 0 -p 5 -k 1 --trim3 97 --norc -s 40000000 bleuven_ION_nov2018_for_indexes A.Dube_Cabsoil_p01_c01_trim_110.fastq index_for_2.txt\n",
      "bowtie -t -v 0 -p 5 -k 1 --trim5 100 --norc -s 40000000 bleuven_ION_nov2018_rev_indexes A.Dube_Cabsoil_p01_c01_trim_110.fastq index_rev_2.txt\n",
      "bowtie -t -v 0 -p 5 -k 1 --trim3 70 --trim5 19 --norc -m 1 -s 40000000 bleuven_ION_nov2018_strain_indexes A.Dube_Cabsoil_p01_c01_trim_110.fastq strain_bc_2.txt\n"
     ]
    }
   ],
   "source": [
    "index_for_align_param = 'bowtie -t -v 0 -p 5 -k 1 --trim3 97 --norc -s 40000000 bleuven_ION_nov2018_for_indexes '\n",
    "# parameters for forward index alignments. Based on optimization work done in BAR-SEQ notebook\n",
    "index_for_align_output = 'index_for_2.txt'\n",
    "# alignment output filepath as string\n",
    "index_for_align_call = index_for_align_param + 'A.Dube_Cabsoil_p01_c01_trim_110.fastq' + ' ' + index_for_align_output\n",
    "\n",
    "print index_for_align_call\n",
    "\n",
    "# alignement call as a string\n",
    "subprocess.check_output(index_for_align_call, shell = True)\n",
    "# performs the alignment\n",
    "\n",
    "\n",
    "\n",
    "index_rev_align_param = 'bowtie -t -v 0 -p 5 -k 1 --trim5 100 --norc -s 40000000 bleuven_ION_nov2018_rev_indexes '\n",
    "# parameters for reverse index alignments. Based on optimization work done in BAR-SEQ notebook\n",
    "index_rev_align_output = 'index_rev_2.txt'\n",
    "# alignment output filepath as string\n",
    "index_rev_align_call = index_rev_align_param + 'A.Dube_Cabsoil_p01_c01_trim_110.fastq' + ' ' + index_rev_align_output\n",
    "# alignement call as a string\n",
    "subprocess.check_output(index_rev_align_call, shell = True)\n",
    "# performs the alignment\n",
    "\n",
    "print index_rev_align_call\n",
    "\n",
    "\n",
    "\n",
    "guide_align_param = 'bowtie -t -v 0 -p 5 -k 1 --trim3 70 --trim5 19 --norc -m 1 -s 40000000 bleuven_ION_nov2018_strain_indexes '\n",
    "# parameters for perfect guide alignments. Based on optimization work done in BAR-SEQ notebook\n",
    "bc_align_output = 'strain_bc_2.txt'\n",
    "# alignment output filepath as string\n",
    "bc_align_call = guide_align_param + 'A.Dube_Cabsoil_p01_c01_trim_110.fastq' + ' ' + bc_align_output\n",
    "# alignement call as a string\n",
    "subprocess.check_output(bc_align_call, shell = True)\n",
    "# performs the alignment\n",
    "\n",
    "print bc_align_call"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11520485\n",
      "10376682\n",
      "9957479\n",
      "10707082\n"
     ]
    }
   ],
   "source": [
    "bc_align_output = 'strain_bc_2.txt'\n",
    "index_for_align_output = 'index_for_2.txt'\n",
    "index_rev_align_output = 'index_rev_2.txt'\n",
    "\n",
    "index_found_for = {}\n",
    "index_found_rev = {}\n",
    "# empty containers that will hold    read_name: index    data pairs for the forward and reverse reads that\n",
    "# had valid alignments\n",
    "\n",
    "\n",
    "with open(index_for_align_output, 'r') as for_index_positives:\n",
    "    for line in for_index_positives:\n",
    "        # opens P1 index alignment pack file and loops through lines\n",
    "\n",
    "        line = line.split('\\t')\n",
    "        # split line to extract data\n",
    "        read_name = line[0]\n",
    "        # Only keep the cluster coordinates and run associated info, not the illumina annotations in the read \n",
    "        # name. See http://support.illumina.com/content/dam/illumina-support/help/BaseSpaceHelp_v2/Content/Vault/\n",
    "        #           Informatics/Sequencing_Analysis/BS/swSEQ_mBS_FASTQFiles.htm\n",
    "        # For info on illumina read name convention\n",
    "        index_found_for[read_name] = line[2]\n",
    "        # add read P1 index to dict\n",
    "\n",
    "print len(index_found_for.keys())\n",
    "# log entry: number of reads with valid P1 index alignments\n",
    "\n",
    "with open(index_rev_align_output, 'r') as rev_index_positives:\n",
    "    for line in rev_index_positives:\n",
    "        # opens P2 index alignment pack file and loops through lines\n",
    "\n",
    "        line = line.split('\\t')\n",
    "        # split line to extract data\n",
    "        read_name = line[0]\n",
    "        # Only keep the cluster coordinates and run associated info\n",
    "        index_found_rev[read_name] = line[2]\n",
    "        # add read P2 index to dict\n",
    "\n",
    "print len(index_found_rev.keys())\n",
    "# log entry: number of reads with valid P2 index alignments\n",
    "\n",
    "\n",
    "both_indexes = set(index_found_for.keys()) & set(index_found_rev.keys())\n",
    "print len(both_indexes)\n",
    "# Use the set data format to quickly id reads with valid alignments for both P1 and P2 indexes, and stores the\n",
    "# taht number for the read pack in the log\n",
    "\n",
    "del both_indexes\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "barcode_match = {}\n",
    "# empty containers that will hold    read_name: guide    data pairs for the forward reads that had alignments\n",
    "# with 0 mismatches at the right location in the reference sequence set\n",
    "\n",
    "with open(bc_align_output, 'r') as matches:\n",
    "    for line in matches:\n",
    "        # opens gRNA alignment pack file and loops through lines\n",
    "\n",
    "        line = line.split('\\t')\n",
    "        # split line to extract data\n",
    "        read_name = line[0]\n",
    "        # Only keep the cluster coordinates and run associated info\n",
    "        guide = line[2]\n",
    "        # the guide for which a match was found\n",
    "\n",
    "        if int(line[3]) in range(0, 5):\n",
    "            # the alignment is at the right location in the guide sequence\n",
    "\n",
    "            barcode_match[read_name] = guide\n",
    "            # add read:guide to dict\n",
    "\n",
    "print len(barcode_match.keys())\n",
    "# adds the number of perfect matches found in the pack of reads\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8992667\n"
     ]
    }
   ],
   "source": [
    "indexes_and_guides = set(index_found_for.keys()) & set(index_found_rev.keys()) & set(barcode_match.keys())\n",
    "print len(indexes_and_guides)\n",
    "# uses sets in python to find the set of reads for which the P1, P2 and gRNA alignments were good"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2256\n"
     ]
    }
   ],
   "source": [
    "for read in indexes_and_guides:\n",
    "    # loops through the reads tht passed the filter\n",
    "\n",
    "    for_i = index_found_for[read]\n",
    "    # gets the P1 index of the read\n",
    "\n",
    "    rev_i = index_found_rev[read]\n",
    "    # gets the P2 index of the read\n",
    "    \n",
    "    if tuple((for_i, rev_i)) in correct_libs:\n",
    "\n",
    "        lib = lib_index_by_tuple[tuple((for_i, rev_i))]\n",
    "        # gets the sample to which the read should be assigned based on P1, P2 combination\n",
    "\n",
    "        guide = barcode_match[read]\n",
    "        # gets the guide associated with the read\n",
    "\n",
    "        guide_counts_counter_dict[lib][guide] += 1\n",
    "        # a count for the guide detected in its specific sample\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        chimera += 1 \n",
    "\n",
    "\n",
    "\n",
    "del index_found_for\n",
    "del index_found_rev\n",
    "del barcode_match\n",
    "\n",
    "print chimera"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100\n",
      "64107.0\n",
      "1371055.0\n",
      "39803250.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x7f73ed9cf150>]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZMAAAD8CAYAAACyyUlaAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3Xt4X1Wd7/H3N0lzvzfpNS0ttBQK\nCtRIy6AOgkJBxzIOKuoMFSvVUUYHdRScM4dzFJ+Dz/HxwozDORxAijhUBhnpOCBTiw4gtjYtcmsp\nDW1pE9rmfr8n3/PHb6WEkgvNL83v9nk9T578fmtf1tpsnny611p7b3N3REREopEW6waIiEjiU5iI\niEjUFCYiIhI1hYmIiERNYSIiIlFTmIiISNQUJiIiEjWFiYiIRE1hIiIiUcuIdQOmS1lZmS9atCjW\nzRARSSg7duxocPfyidZLmTBZtGgRVVVVsW6GiEhCMbNX38p6E3ZzmdndZlZnZi+MsuwrZuZmVha+\nm5ndZmbVZvacma0Yse5aM9sbftaOKH+HmT0ftrnNzCyUl5rZ5rD+ZjMrmagOERGJjbcyZnIPsPr4\nQjNbAFwKHBxRfDmwNPysB24P65YCNwMrgfOBm4fDIaxz3Yjthuu6Edji7kuBLeH7mHWIiEjsTBgm\n7v4E0DTKou8DXwNGPnZ4DXCvR2wFis1sLnAZsNndm9y9GdgMrA7LCt19q0ceX3wvcOWIfW0Inzcc\nVz5aHSIiEiOTms1lZmuAWnd/9rhF84FDI77XhLLxymtGKQeY7e6Hw+cjwOwJ6hARkRg54QF4M8sF\nvkGki2tauLub2Qm/eMXM1hPpCmPhwoVT3i4REYmYzJXJacBi4FkzOwBUADvNbA5QCywYsW5FKBuv\nvGKUcoCjw91X4XddKB9rX2/i7ne4e6W7V5aXTzizTUREJumEw8Tdn3f3We6+yN0XEelmWuHuR4BN\nwDVhxtUqoDV0VT0GXGpmJWHg/VLgsbCszcxWhVlc1wAPh6o2AcOzvtYeVz5aHSIiEiNvZWrw/cDv\ngWVmVmNm68ZZ/RFgH1AN/D/g8wDu3gR8C9gefr4Zygjr3Bm2eQV4NJTfCrzfzPYC7wvfx6xDRETe\n7Ae/fpmnqxtOej2WKu+Ar6ysdN20KCKppLmzjxW3bOaG953OFy9ZOql9mNkOd6+caD09m0tEJElt\n29+IO/zJaTNPel0KExGRJPX0K43kZqbz9orik16XwkREJEk9/Uoj71xUSmbGyf9TrzAREUlCdW09\nVNd1TEsXFyhMRESS0u/3NQLwJ6eVTUt9ChMRkST0dHUjhdkZLJ9XOC31KUxERJLQ0/saWHXqTNLT\nbFrqU5iIiCSZQ01dHGrqnrbxElCYiIgknd+/EsZLlkzPeAkoTEREks7vXmmgLD+TpbPyp61OhYmI\nSBJxd55+pZELTisjvAV9WihMRESSyN66Durbe6d1vAQUJiIiSeXJvZEnBL9rGsdLQGEiIpJUntpb\nz+KyPBaU5k5rvQoTEZEk0TcwxLb9TdN+VQIKExGRpLHzYDNdfYO8a6nCREREJunJvfWkpxkXTPPg\nOyhMRESSxlN7Gzh3QTGF2TOmvW6FiYhIEmjp6uO52taYjJeAwkREJCk8/UrkFb3vjsF4CShMRESS\nwpN7G8jPyuCcBSf/Fb2jmTBMzOxuM6szsxdGlP1vM3vJzJ4zs38zs+IRy24ys2oz22Nml40oXx3K\nqs3sxhHli81sWyj/mZllhvKs8L06LF80UR0iIqnI3Xlybz0XnDaTGemxuUZ4K7XeA6w+rmwzcLa7\nvx14GbgJwMyWA1cDZ4Vt/tnM0s0sHfgRcDmwHPh4WBfgO8D33X0J0AysC+XrgOZQ/v2w3ph1nOBx\ni4gkjbr2Xmqau7ng1OmfxTVswjBx9yeApuPK/tPdB8LXrUBF+LwG2Ojuve6+H6gGzg8/1e6+z937\ngI3AGos8hexi4MGw/QbgyhH72hA+PwhcEtYfqw4RkZTU2NEHwNyi7Ji1YSquhz4NPBo+zwcOjVhW\nE8rGKp8JtIwIpuHyN+wrLG8N64+1LxGRlNTe0w9AQQymBA+LKkzM7O+BAeCnU9OcqWVm682sysyq\n6uvrY90cEZGToq0n8u/xwpyMmLVh0mFiZp8CPgh80t09FNcCC0asVhHKxipvBIrNLOO48jfsKywv\nCuuPta83cfc73L3S3SvLy8sncZQiIvGvrTtyZRKLmxWHTSpMzGw18DXgQ+7eNWLRJuDqMBNrMbAU\n+AOwHVgaZm5lEhlA3xRC6DfAVWH7tcDDI/a1Nny+Cng8rD9WHSIiKen1bq7YXZlMWLOZ3Q9cBJSZ\nWQ1wM5HZW1nA5vAmr63u/jl3f9HMHgB2Een++oK7D4b9XA88BqQDd7v7i6GKrwMbzewW4BngrlB+\nF/ATM6smMgHgaoDx6hARSUXD3VyxHDOx13uokltlZaVXVVXFuhkiIlPull/u4qfbDrL7W8ffxRE9\nM9vh7pUTrac74EVEElx7z0BMu7hAYSIikvDaevopzIldFxcoTEREEl57zwCFujIREZFotPX0x3Tw\nHRQmIiIJr61b3VwiIhIldXOJiEhU3F0D8CIiEp2e/iH6B11Tg0VEZPKGH6USy+dygcJERCShtQ2H\nibq5RERkslq7h5/LpW4uERGZJHVziYhI1IafGFwUwxdjgcJERCShDb8YS3fAi4jIpLUPv7JXYSIi\nIpPV1tPPjHQje0Zs/5wrTEREElhbd+Qhj+GttzGjMBERSWDx8FwuUJiIiCS0eHguFyhMREQSWqSb\nS1cmIiIShUg3l65MREQkCm09/YkRJmZ2t5nVmdkLI8pKzWyzme0Nv0tCuZnZbWZWbWbPmdmKEdus\nDevvNbO1I8rfYWbPh21uszAlYTJ1iIikmvaegYTp5roHWH1c2Y3AFndfCmwJ3wEuB5aGn/XA7RAJ\nBuBmYCVwPnDzcDiEda4bsd3qydQhIpJq+geH6OobTIwBeHd/Amg6rngNsCF83gBcOaL8Xo/YChSb\n2VzgMmCzuze5ezOwGVgdlhW6+1Z3d+De4/Z1InWIiKSU1+9+T4wrk9HMdvfD4fMRYHb4PB84NGK9\nmlA2XnnNKOWTqeNNzGy9mVWZWVV9ff1bPDQRkcQw/MTgWD+XC6ZgAD5cUfgUtGXK63D3O9y90t0r\ny8vLT0LLRERipy28yyQhurnGcHS4ayn8rgvltcCCEetVhLLxyitGKZ9MHSIiKeXYWxYTuJtrEzA8\nI2st8PCI8mvCjKtVQGvoqnoMuNTMSsLA+6XAY2FZm5mtCrO4rjluXydSh4hISomnbq4J48zM7gcu\nAsrMrIbIrKxbgQfMbB3wKvDRsPojwBVANdAFXAvg7k1m9i1ge1jvm+4+PKj/eSIzxnKAR8MPJ1qH\niEiqeb2bK/ZXJhO2wN0/PsaiS0ZZ14EvjLGfu4G7RymvAs4epbzxROsQEUklx7q5EnjMREREYqyt\nZwAzyM+M/ZWJwkREJEG1dfeTn5VBWlps32UCChMRkYQVL8/lAoWJiEjCau8ZiIvxElCYiIgkrHh5\nlwkoTEREElZbnLzLBBQmIiIJq72nPy7uMQGFiYhIwmrr1gC8iIhEYWjIae8diIvncoHCREQkIXX2\nDeAeH3e/g8JERCQhtYUXY2k2l4iITFpb9/Dj53VlIiIik3S4tRuA2UXZMW5JhMJERCQB1TRHwqSi\nJCfGLYlQmIiIJKCa5m6yMtIoz8+KdVMAhYmISEI61NRFRUkOkZfUxp7CREQkAdU0d1NRkhvrZhyj\nMBERSUCHmrtYUBof4yWgMBERSTjtPf20dPXrykRERCZveCbXAoWJiIhMVrxNC4Yow8TMbjCzF83s\nBTO738yyzWyxmW0zs2oz+5mZZYZ1s8L36rB80Yj93BTK95jZZSPKV4eyajO7cUT5qHWIiKSCQ01d\nQJKEiZnNB74IVLr72UA6cDXwHeD77r4EaAbWhU3WAc2h/PthPcxsedjuLGA18M9mlm5m6cCPgMuB\n5cDHw7qMU4eISNKrae4mNzOd0rz4+Xd0tN1cGUCOmWUAucBh4GLgwbB8A3Bl+LwmfCcsv8QiE6TX\nABvdvdfd9wPVwPnhp9rd97l7H7ARWBO2GasOEZGkV9McX/eYQBRh4u61wHeBg0RCpBXYAbS4+0BY\nrQaYHz7PBw6FbQfC+jNHlh+3zVjlM8epQ0Qk6R1q7o6rwXeIrpurhMhVxWJgHpBHpJsqbpjZejOr\nMrOq+vr6WDdHRGRKDF+ZxJNourneB+x393p37wceAi4EikO3F0AFUBs+1wILAMLyIqBxZPlx24xV\n3jhOHW/g7ne4e6W7V5aXl0dxqCIi8aG1q5/2ngEWlCbJlQmR7q1VZpYbxjEuAXYBvwGuCuusBR4O\nnzeF74Tlj7u7h/Krw2yvxcBS4A/AdmBpmLmVSWSQflPYZqw6RESS2qHm+JvJBdGNmWwjMgi+E3g+\n7OsO4OvAl82smsj4xl1hk7uAmaH8y8CNYT8vAg8QCaJfAV9w98EwJnI98BiwG3ggrMs4dYiIJLWa\nY2ESX1cmUb3v0d1vBm4+rngfkZlYx6/bA3xkjP18G/j2KOWPAI+MUj5qHSIiyS4e734H3QEvIpJQ\napq7KcjKoDAnPt79PkxhIiKSQA41dVFRmhtX95iAwkREJKFE3mMSX4PvoDAREUkY7s6hOLzHBBQm\nIiIJo7mrn66+wbgbfAeFiYhIwjgYh08LHqYwERFJEL94ppaMNONtFUWxbsqbKExERBJAfXsvG7cf\n5Mrz5jO3SFcmIiIyCXc9tZ/egSH++qLTYt2UUSlMRETiXGtXP/dtfZUPvG0up5Xnx7o5o1KYiIjE\nuXuePkBH7wBfeO+SWDdlTAoTEZE41tE7wI+f3s/7zpzNmXMLY92cMSlMRETi1OCQ87UHn6Wlq5/r\nL47fqxJQmIiIxCV35+//7Xkeef4I//DB5Zy7oDjWTRqXwkREJM64O7c++hIbtx/iixcvYd27Fse6\nSROKr2cYi4ikuIHBIb79yG5+/LsDrL3gFG54/+mxbtJbojAREYkTTZ19XP8vO3n6lUauvXAR//CB\n5XH3qPmxKExERGKosaOXPUfb2XOknTuf3E99Ry/f/cg5XPWOilg37YQoTEREppm781R1A7dt2cv2\nA83HyheW5vKvn72Ac+J8sH00ChMRkWn04mut/MMvXmDnwRbmFmXzd5ct4+0VRSybXUB5QVbCdGsd\nT2EiIjJNhoacL238Iy1d/dxy5dl8pLKCrIz0WDdrSkQ1NdjMis3sQTN7ycx2m9kFZlZqZpvNbG/4\nXRLWNTO7zcyqzew5M1sxYj9rw/p7zWztiPJ3mNnzYZvbLET2WHWIiMSzR184QnVdBzf/2XL+ctUp\nSRMkEP19Jj8EfuXuZwDnALuBG4Et7r4U2BK+A1wOLA0/64HbIRIMwM3ASuB84OYR4XA7cN2I7VaH\n8rHqEBGJS0NDzj8+vpdTy/O44m1zY92cKTfpMDGzIuA9wF0A7t7n7i3AGmBDWG0DcGX4vAa41yO2\nAsVmNhe4DNjs7k3u3gxsBlaHZYXuvtXdHbj3uH2NVoeISFzavPsoLx1p5/r3LiE9LTHHRcYTzZXJ\nYqAe+LGZPWNmd5pZHjDb3Q+HdY4As8Pn+cChEdvXhLLxymtGKWecOkRE4o575KrklJm5fOicebFu\nzkkRTZhkACuA2939PKCT47qbwhWFR1HHhMarw8zWm1mVmVXV19efzGaIiIzpt3vqeaG2jc9fdBoZ\n6cn5FKtojqoGqHH3beH7g0TC5WjooiL8rgvLa4EFI7avCGXjlVeMUs44dbyBu9/h7pXuXlleXj6p\ngxQRmayXjrRx48+f43P37WB+cQ5/fl5i3Yh4IiYdJu5+BDhkZstC0SXALmATMDwjay3wcPi8Cbgm\nzOpaBbSGrqrHgEvNrCQMvF8KPBaWtZnZqjCL65rj9jVaHSIiMdfQ0ctnNlSx+gdP8os/1vLhFRX8\ny3UrycxIzqsSiP4+k78BfmpmmcA+4FoiAfWAma0DXgU+GtZ9BLgCqAa6wrq4e5OZfQvYHtb7prs3\nhc+fB+4BcoBHww/ArWPUISISU0/ureeGnz1LW08/f3fZMj65ciHFuZmxbtZJZ5Ehh+RXWVnpVVVV\nsW6GiCSZ1u5+XqxtZW9dB88eauGhZ2pZOiuff/zEeZwxJ37fjPhWmdkOd6+caD3dAS8iMkk7Xm3m\n2h//gbaeAQAKsjP4q1Wn8I0rziQnM3luSHwrFCYiIpPwdHUDn7m3ilkFWfzjJ1Zw5pzEfrZWtBQm\nIiIn6PGXjvK5+3ayaGYu961byazC7Fg3KeYUJiIiJ2DrvkY+95OdnD4nn3s/vZLSvOQfXH8rFCYi\nIm9RdV076++tYkFpDj9dt4qi3BmxblLcSN5JzyIiU6i+vZdP/Xg7mRlp3HPt+QqS4+jKRERkDDXN\nXew+3M6eI238+7OHaezoY+P6VSwozY110+KOwkREUpq7U9/RS3NnPy1dfRxp62Hrvkaeqm7gUFP3\nsfUWlubyo0+el5Cv1J0OChMRSTlH23r4+c4anjnYwjMHW2jo6H3D8oKsDC44bSafedepvK2iiKWz\n8inIVrfWeBQmIpIyWrv6uf2/XuHHv9tP78AQp5bl8Z7Ty3jb/CJmFWRTnDuD0rxMls7KT9qn+54s\nChMRSWoDg0NsP9DMYy8e4aGdNbT3DnDlufO54X2ns3Cmxj6misJERJLS4JBz55P7+L9P7KOps4+s\njDQuOXMWf3PxUs6cm/jPzIo3ChMRSTo1zV185YFn2ba/iYuWlfPRygVctKyc3Ez9yTtZ9F9WRBKS\nu9PQ0UdNcxeHmrupa+uhuauPps5+fvnsazjw3Y+cw1+smJ+yz8uaTgoTEUkYT+1t4PGX6njxtVZ2\nHW6jPTytd1h6mlGSm8mKU0q45cqzdT/INFKYiEjc6+kf5Jb/2MV9Ww+SMyOdM+YW8KFz5rFkVj4L\nSnJZUJrLnMJsCnMydBUSIwoTEYlre4608zf37+Tlox1c9+7FfPWyZWRlpNa7QhKBwkRE4taeI+1c\ndfvTZM1IY8Onz+dPTy+PdZNkDAoTEYlLdW09fPqe7eRkpvNvX7iQ+cU5sW6SjENhIiJxp7N3gE9v\n2E5zVx8PfPYCBUkCUJiISMzVt/ey49Umapq7qW/vZdv+Jna91sadays5e35RrJsnb0HUYWJm6UAV\nUOvuHzSzxcBGYCawA/grd+8zsyzgXuAdQCPwMXc/EPZxE7AOGAS+6O6PhfLVwA+BdOBOd781lI9a\nR7THIiInX//gEC8fbef5mlaerWlh2/4m9tV3HluemZHG7MIsvvMXb+fiM2bHsKVyIqbiyuRLwG5g\n+PkE3wG+7+4bzez/EAmJ28PvZndfYmZXh/U+ZmbLgauBs4B5wK/N7PSwrx8B7wdqgO1mtsndd41T\nh4jESO/AIDsONFPf0UtDRx/NnX20dPfR0tVPa3c/jR19NHb20tTZR/+gA1CYncE7F5XyscoFvHNx\nKaeV5Wt6b4KKKkzMrAL4APBt4MsW+T/gYuATYZUNwP8g8od+TfgM8CDwT2H9NcBGd+8F9ptZNXB+\nWK/a3feFujYCa8xs9zh1iEgM/GF/Ezc99ByvjLjCSDMozs2kKGcGRTkzmFuUzVnzCikryOKMOQWc\nU1HMKTNzFRxJItorkx8AXwMKwveZQIu7D9+WWgPMD5/nA4cA3H3AzFrD+vOBrSP2OXKbQ8eVr5yg\nDhGZBp29A9S391Lf0ctDO2u5/w8HqSjJ4UefWMGyOQWU5WdSmD2DtDQFRaqYdJiY2QeBOnffYWYX\nTV2Tpo6ZrQfWAyxcuDDGrRFJbPXtvTz8x1p+vrOW3YfbjpWnGVz37sXc8P7T9SDFFBbNmb8Q+JCZ\nXQFkExkz+SFQbGYZ4cqhAqgN69cCC4AaM8sAiogMxA+XDxu5zWjljePU8QbufgdwB0BlZaVHcawi\nKWloyHlibz33bT3Ib/bUMTjknFNRxFcvPZ25RTmUF2RxankeFSV6Blaqm3SYuPtNwE0A4crkq+7+\nSTP7V+AqIrOt1gIPh002he+/D8sfd3c3s03Av5jZ94gMwC8F/gAYsDTM3KolMkj/ibDNb8aoQ0Qm\nMDTktHT309U3QN/AEH2DQwwMOu4w5E5TZx8Hm7o42NTF5l1HOdjURVl+Jte9+1Suesd8lswqmLgS\nSTkn45r068BGM7sFeAa4K5TfBfwkDLA3EQkH3P1FM3sA2AUMAF9w90EAM7seeIzI1OC73f3FCeoQ\nSWmNHb088vxhWrr66R0Yoqd/kKbOPuo7eqlv76Wxs4+mzj4Ghya+UM+ekca5C4r5u8uWcdlZc8jM\n0GtsZWzmnhq9P5WVlV5VVRXrZohMucEhZ29dOxuePsBDO2vpHRgCImMZ2TPSKcnNpKwgi/L8TMry\ns5iZn8nMvCzyszLIzEgjMyONjDQjzQwzKM6dwYLSXMrzszTTSjCzHe5eOdF6Gi0TiSM9/YPHriA6\negbo6I3co9HQ0Ud9ey/NXX109g7S1TdAR+8AdW2RGVWDQ05WRhofXlHBtRcu4tSyPDLSdSUh00dh\nInKS9A5EguFoWy+t3X209wzQ3jPA0bYeDjR28WpjJ02dfQwNOYPudPUNvullTyPlZ2VQmpdJXlYG\neZmRK45lswuYXZjNvOIcVp89h9K8zGk8QpHXKUxEotA7MEhf6FYaGHR2vNrM43vq+K899dS2dI+6\nTZpBRUkup8zMZUl5PmlpRroZOZnplBdkUV6Qxcy8TAqyZ5CflUFhTgZl+Vlkz9A7PCR+KUxExuHu\ntPUM0NTZR0MYxD7S2sOuw208X9PK3rp2jh/Lzs1M511LyvjYOxcwqyCLWYVZlIYxisLsDIpzMzWY\nLUlHYSIp7ZHnD/OT379Ke28/Xb2DdPcPMuTOkEcGtlu7+0ed+VSWn8nbK4q57KzZFObMOFZ+xpxC\n3rm4RG8ClJSjMJGU1NU3wDf/fRcbtx/i1PI8Fs/MI7csg+yMNNLTIrOa0swozp1BSW4mpXmZlBdk\nUZb/ejeUZjqJvE5hIkmtdyByn0VjR6SbavjJtT/bfoh9DZ389UWn8eX3n84MzXwSiYrCRBLK4JDT\n1t1PS3c/vQODuIM7tHb381pLN4dbuznY1MX+hk72N3TS0DH6a27mFWVz37qVXLikbJqPQCQ5KUwk\nprr6BnitpYfXWro50tbD0dYe6tp76R0YZGDQ6R9ymjv7qGuPlLd29zPRfbZl+VmcWpbHJWfMZn5J\nzogb9SI37ZUVZJGXma5uKpEppDCRKdc/OMSR1h6OtkUC4GhbD4dbI4FR19ZLe2/kZry27gFau/vf\ntH1J7gxyZqSTnh6ZMluSl8nisjxWLp5JaV4mxbmR92Nkz0jHADPIz5rB/JIc5hZlawqtSAwoTOSE\nDAwOUdfey5BHHgzY2NnHS4fbeOlIO9V1HRxo7OS1lu43TZfNzEhjblE2swuzqSjJoSCrgPzsDOYU\nZTOvKBICc4tymFWo+ylEEpHCRN6yPx5q4Yaf/ZH9DZ1vWpaXmc6SWfmsWFjCn583n4qSHGYXZjOr\nIJtZhZr9JJLsFCYyoYHBIf75t6/wwy17mV2QxTfXnHWsi6kwZwZnzimkoiRHb9UTSWEKE8Hd6egd\noKGjj8aOXhrCNNoDDZ28XNfBS4fbqGvvZc258/jmmrMpGnGTnogIKExSQkfvAM8cbOZAYxc1zV28\n1tJDU2cvzZ3DT6TtPfbY8pGyMtJYMiufC5eUceny2Vz+trkxaL2IJAKFSRJyd3YfbudXLxzmqeoG\nnq1pPfZIkBnpxrziyHTZuUXZnDG3IDJdNrzrYnga7fDndHVdichboDBJcENDzu4jbcfedbG/oYv/\neO41XqnvJM3g7RXFfPY9p3LBaTNZOquAWQVZGtsQkSmnMElgDR29fPmBZ3ni5fpjZWZw/qJSrr1w\nMZefPYeZ+VkxbKGIpAqFSYL6/SuNfGnjM7R09/P3V5zJilOKKc2LdFcVZGuAXESml8IkgbhHXr50\n39ZX2fTsayyamcc9157P8nmFsW6aiKQ4hUkcO9rWwwu1rRxs6uJgUxe/f6WRl460U5CVwbUXLubL\n7z+dvCydQhGJPf0lihNDQ86+hk6ePdRC1atNbN3X9IY7zXMz01k2p4BbP/w2PnTuPHIzdepEJH5M\n+i+SmS0A7gVmAw7c4e4/NLNS4GfAIuAA8FF3b7bIszR+CFwBdAGfcvedYV9rgf8Wdn2Lu28I5e8A\n7gFygEeAL7m7j1XHZI8lltp6+vlfj+zml88dpr1nAICC7AxWLi7lkysXct7CYk6ZmafHkYhIXIvm\nn7cDwFfcfaeZFQA7zGwz8Clgi7vfamY3AjcCXwcuB5aGn5XA7cDKEAw3A5VEQmmHmW0K4XA7cB2w\njUiYrAYeDfscrY6EsnVfI1954FmOtPXw4fPm887FpZy7oJjTyvN1f4eIJJRJh4m7HwYOh8/tZrYb\nmA+sAS4Kq20AfkvkD/0a4F53d2CrmRWb2dyw7mZ3bwIIgbTazH4LFLr71lB+L3AlkTAZq46E0Nrd\nzw9+/TL3PH2ARTPzePBzF3DewpJYN0tEZNKmpOPdzBYB5xG5gpgdggbgCJFuMIgEzaERm9WEsvHK\na0YpZ5w6jm/XemA9wMKFC0/wqKbewOAQG7cf4nubX6a5q49PrlzIN644U+MfIpLwov4rZmb5wM+B\nv3X3tpH9+mF8Y4L34kVnvDrc/Q7gDoDKysqT2o6JuDt/edc2tu5r4vzFpfz3Dy7n7PlFsWySiMiU\niSpMzGwGkSD5qbs/FIqPmtlcdz8curHqQnktsGDE5hWhrJbXu6yGy38byitGWX+8OuLWf+46ytZ9\nTfy3D5zJunct1mC6iCSVtMluGGZn3QXsdvfvjVi0CVgbPq8FHh5Rfo1FrAJaQ1fVY8ClZlZiZiXA\npcBjYVmbma0KdV1z3L5GqyMuuTu3bdnLopm5fOpPFilIRCTpRHNlciHwV8DzZvbHUPYN4FbgATNb\nB7wKfDQse4TItOBqIlODrwVw9yYz+xawPaz3zeHBeODzvD41+NHwwzh1xKVf767jxdfa+O5HziEj\nfdL5LSIStywyuSr5VVZWelVZ3JWEAAAGN0lEQVRV1bTX6+782T89RXvPAFu+/KcKExFJKGa2w90r\nJ1pPf9lOsi2763ihto3r37tEQSIiSUtzUk+C+/9wkJcOt9Hc1c/2A00sLM3lz8+bP/GGIiIJSmEy\nxfbVd3DTQ8+Tn5XBzPxM5hRl89VLl+mqRESSmsJkim3ZHZml/Ku/fTcVJbkxbo2IyPTQP5en2K93\nH+WMOQUKEhFJKQqTKdTa1U/Vq81ccuasWDdFRGRaKUym0G9frmNwyLnkzFEfFSYikrQUJlNoy+46\nyvIzObeiONZNERGZVgqTKdI/OMRv99Tx3mWzSNO7SEQkxShMpkjVgWbaegbUxSUiKUlhMkW27D5K\nZnoa715aFuumiIhMO4XJFHB3trxUx6rTZpKXpVt3RCT16C9fFHr6B3loZy13/24/+xs6+ex7To11\nk0REYkJhcgLcnbr2Xp7a28ATe+v5r5fraenq56x5hXzvo+dw5bl6/paIpCaFyQR+s6eOb/1yF61d\n/bR09zM4FHlkf1l+Ju9dNouPvXMBKxeX6oVXIpLSFCYTKM6ZwZlzCynOmUFJbialeZmcv7iU5XML\nNQVYRCRQmEzgvIUl/OgTJbFuhohIXNNsLhERiZrCREREoqYwERGRqClMREQkagkdJma22sz2mFm1\nmd0Y6/aIiKSqhA0TM0sHfgRcDiwHPm5my2PbKhGR1JSwYQKcD1S7+z537wM2Amti3CYRkZSUyGEy\nHzg04ntNKBMRkWmW1Dctmtl6YH342mFmeya5qzKgYWpalVBS8bhT8ZghNY87FY8ZTvy4T3krKyVy\nmNQCC0Z8rwhlx7j7HcAd0VZkZlXuXhntfhJNKh53Kh4zpOZxp+Ixw8k77kTu5toOLDWzxWaWCVwN\nbIpxm0REUlLCXpm4+4CZXQ88BqQDd7v7izFulohISkrYMAFw90eAR6ahqqi7yhJUKh53Kh4zpOZx\np+Ixw0k6bnP3k7FfERFJIYk8ZiIiInFCYTKBVHhki5ktMLPfmNkuM3vRzL4UykvNbLOZ7Q2/k/LF\nLmaWbmbPmNkvw/fFZrYtnPOfhQkeScPMis3sQTN7ycx2m9kFqXCuzeyG8P/3C2Z2v5llJ+O5NrO7\nzazOzF4YUTbq+bWI28LxP2dmKyZbr8JkHCn0yJYB4CvuvhxYBXwhHOeNwBZ3XwpsCd+T0ZeA3SO+\nfwf4vrsvAZqBdTFp1cnzQ+BX7n4GcA6RY0/qc21m84EvApXufjaRSTtXk5zn+h5g9XFlY53fy4Gl\n4Wc9cPtkK1WYjC8lHtni7ofdfWf43E7kj8t8Ise6Iay2AbgyNi08ecysAvgAcGf4bsDFwINhlaQ6\nbjMrAt4D3AXg7n3u3kIKnGsiE45yzCwDyAUOk4Tn2t2fAJqOKx7r/K4B7vWIrUCxmc2dTL0Kk/Gl\n3CNbzGwRcB6wDZjt7ofDoiPA7Bg162T6AfA1YCh8nwm0uPtA+J5s53wxUA/8OHTt3WlmeST5uXb3\nWuC7wEEiIdIK7CC5z/VIY53fKfsbpzCRY8wsH/g58Lfu3jZymUem/SXV1D8z+yBQ5+47Yt2WaZQB\nrABud/fzgE6O69JK0nNdQuRf4YuBeUAeb+4KSgkn6/wqTMY34SNbkoWZzSASJD9194dC8dHhS97w\nuy5W7TtJLgQ+ZGYHiHRhXkxkPKE4dIVA8p3zGqDG3beF7w8SCZdkP9fvA/a7e7279wMPETn/yXyu\nRxrr/E7Z3ziFyfhS4pEtYZzgLmC3u39vxKJNwNrweS3w8HS37WRy95vcvcLdFxE5t4+7+yeB3wBX\nhdWS6rjd/QhwyMyWhaJLgF0k+bkm0r21ysxyw//vw8edtOf6OGOd303ANWFW1yqgdUR32AnRTYsT\nMLMriPSrDz+y5dsxbtKUM7N3AU8Cz/P62ME3iIybPAAsBF4FPuruxw/sJQUzuwj4qrt/0MxOJXKl\nUgo8A/ylu/fGsn1TyczOJTLhIBPYB1xL5B+WSX2uzex/Ah8jMnvxGeAzRMYHkupcm9n9wEVEng58\nFLgZ+AWjnN8QrP9EpMuvC7jW3asmVa/CREREoqVuLhERiZrCREREoqYwERGRqClMREQkagoTERGJ\nmsJERESipjAREZGoKUxERCRq/x+Q0Cps2rvYBQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 600x400 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "counts_DF = pd.DataFrame.from_dict(guide_counts_counter_dict)\n",
    "\n",
    "counts_DF.head(10)\n",
    "\n",
    "\n",
    "print len(counts_DF.columns)\n",
    "\n",
    "print min(counts_DF.sum(axis =0))\n",
    "print max(counts_DF.sum(axis =0))\n",
    "\n",
    "print sum(counts_DF.sum(axis =0))\n",
    "\n",
    "\n",
    "\n",
    "plt.plot(sorted(counts_DF.sum(axis =0)))   \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "counts_DF.head(10)\n",
    "\n",
    "\n",
    "counts_DF.to_csv('./read_count_table_harsh.csv', sep = ',', index_label='Strain')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
